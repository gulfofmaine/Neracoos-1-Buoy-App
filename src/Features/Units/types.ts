/**
 * Types related to unit managent
 */

export enum UnitSystem {
  metric = "Metric",
  imperial = "Imperial"
}

export interface UnitStoreState {
  system: UnitSystem
}

export const initialStoreState: UnitStoreState = {
  system: UnitSystem.imperial
}
