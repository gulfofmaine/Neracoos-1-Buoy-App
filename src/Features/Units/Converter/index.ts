import { data_types } from "./data_types"
import { DataTypeConversion } from "./conversions"

/**
 * Given a known data type, return the DataTypeConversion
 *
 * @param data_type A known data type
 */
export const converter = (data_type: string): DataTypeConversion => {
  return data_types[data_type]
}
