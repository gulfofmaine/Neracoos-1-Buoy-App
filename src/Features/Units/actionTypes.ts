/**
 * Unit system management action types
 */

export const UNIT_SWITCH = "unit/switch"
export type UNIT_SWITCH = typeof UNIT_SWITCH
