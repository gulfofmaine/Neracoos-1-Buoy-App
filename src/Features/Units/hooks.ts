/**
 * Hooks to connect the current unit system
 */
import { useSelector } from "react-redux"

import { StoreState } from "Shared/constants/store"

const currentUnitSystem = (state: StoreState) => state.unit.system

/**
 * Return the current unit system that is connected to Redux updates
 */
export const useUnitSystem = () => useSelector(currentUnitSystem)
