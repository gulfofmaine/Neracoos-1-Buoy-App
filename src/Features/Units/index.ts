export { unitReducer } from "./reducer"

export { UnitSelector } from "./UnitSelector"
