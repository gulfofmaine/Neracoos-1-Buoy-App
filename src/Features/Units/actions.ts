import * as actionTypes from "./actionTypes"
import { UnitSystem } from "./types"

// Action types

export interface UnitSwitch {
  type: typeof actionTypes.UNIT_SWITCH
  system: UnitSystem
}

export type UnitActions = UnitSwitch

// Action creators

/**
 * Switch display to use designated system of units
 * @param system system of  units to switch display into
 */
export const unitSwitch = (system: UnitSystem): UnitSwitch => {
  return {
    system,
    type: actionTypes.UNIT_SWITCH,
  }
}
