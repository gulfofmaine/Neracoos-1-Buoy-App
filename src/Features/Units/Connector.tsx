/**
 * Connect to the current unit system from Redux store
 */
import * as React from "react"

import { UnitSystem } from "Features/Units/types"

import { useUnitSystem } from "./hooks"

interface Props {
  children(props: RenderProps): JSX.Element
}

interface RenderProps {
  unitSystem: UnitSystem
}

/**
 * Connect to the current unit system from Redux store
 */
export const UnitSystemConnector: React.FunctionComponent<Props> = ({ children }) => {
  const unitSystem = useUnitSystem()

  return children({ unitSystem })
}
