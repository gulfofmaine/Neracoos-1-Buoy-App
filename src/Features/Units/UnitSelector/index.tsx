/**
 * Component to select and display currently selected unit type
 */
import * as React from "react"
import { Button, ButtonGroup } from "reactstrap"
import { connect } from "react-redux"
import { bindActionCreators, Dispatch } from "redux"

import { StoreState } from "Shared/constants/store"

import { unitSwitch } from "../actions"
import { UnitSystem } from "../types"

export interface ReduxProps {
  system: UnitSystem
  switchUnits: (system: UnitSystem) => void
}

const mapStateToProps = ({ unit }: StoreState): Pick<ReduxProps, "system"> => {
  return {
    system: unit.system
  }
}

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      switchUnits: unitSwitch
    },
    dispatch
  )

export class UnitSelectorBase extends React.Component<ReduxProps, object> {
  public render(): React.ReactNode {
    return (
      <ButtonGroup>
        {this.unitButton(UnitSystem.metric)}
        {this.unitButton(UnitSystem.imperial)}
        {this.unitButton(UnitSystem.mariners)}
      </ButtonGroup>
    )
  }

  private unitButton(buttonSystem: UnitSystem): React.ReactNode {
    const { system } = this.props

    return (
      <Button
        color="primary"
        size="sm"
        id={buttonSystem}
        active={system === buttonSystem}
        onClick={this.setSystem(buttonSystem)}
      >
        {buttonSystem}
      </Button>
    )
  }

  private setSystem = (system: UnitSystem) => () => {
    this.props.switchUnits(system)
  }
}

export const UnitSelector = connect(
  mapStateToProps,
  mapDispatchToProps
)(UnitSelectorBase)
