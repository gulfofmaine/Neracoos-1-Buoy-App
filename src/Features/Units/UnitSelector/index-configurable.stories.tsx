import React from "react"
import { select } from "@storybook/addon-knobs"
import { action } from "@storybook/addon-actions"

import { UnitSelectorBase } from "./index"
import { UnitSystem } from "Features/Units/types"

export default {
  component: UnitSelectorBase,
  title: "Units|UnitSelector/configurable",
}

export const configurable = () => {
  const options = [UnitSystem.english, UnitSystem.metric]
  const unit = select("Unit System", options, options[0], "unit-system-0")

  return <UnitSelectorBase system={unit} switchUnits={action("switch-units")} />
}

export const english = () => <UnitSelectorBase system={UnitSystem.english} />

export const metric = () => <UnitSelectorBase system={UnitSystem.metric} />
