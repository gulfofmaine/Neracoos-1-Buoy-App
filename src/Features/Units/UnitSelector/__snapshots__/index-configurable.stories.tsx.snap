// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Units/UnitSelector/configurable english smoke-test 1`] = `
<div role="group"
     class="unit-selector btn-group"
>
  <button type="button"
          id="Metric"
          class="btn btn-light btn-sm"
  >
    Metric
  </button>
  <button type="button"
          id="English"
          class="btn btn-primary btn-sm active"
  >
    English
  </button>
</div>
`;

exports[`Units/UnitSelector/configurable metric smoke-test 1`] = `
<div role="group"
     class="unit-selector btn-group"
>
  <button type="button"
          id="Metric"
          class="btn btn-primary btn-sm active"
  >
    Metric
  </button>
  <button type="button"
          id="English"
          class="btn btn-light btn-sm"
  >
    English
  </button>
</div>
`;
